bpf/bpf_lxc.c

```diff
@@ -874,11 +874,7 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_TO_LXC) int tail_ipv6_policy(s
  * Returns true if the packet should pass, false if it should be dropped. */
 static bool check_ip4_fragments(struct iphdr *ip4)
```

Comment:
The idea behind this function is to return true if either:
* The packet is not an IP fragment, or
* The packet is an IP fragment AND:
  * Policy is disabled on ingress, OR
  * Policy is enabled on ingress, but this traffic would be allowed at L3

And to return false if:
* This packet is an IP fragment, Policy is enabled on ingress, and this traffic would not be allowed at L3

In practice, the policy part of this is a bit naïve and simply checks whether ingress policy enforcement is enabled instead, relying on these `#define`s.

These changes as-is will now always drop any IP fragments that are received. When we run with overlays with kube-proxy, we do not expect to receive any fragments. However, if kube-proxy is not in use, or we run in direct routing mode, it is possible to receive fragments and now we will unconditionally drop them *even if there is no policy in place that might require IP fragments to be reassembled*. (For context, we cannot perform L4+ policy on IP fragments, because IP fragments after the first one do not contain the L4 port and therefore we don't know which port to use to apply L4 policy).

To avoid regressing on this, we would need to have another version of `__policy_can_access()` which operates on fragment packets and skips the L4 lookups (the parts protected by `HAVE_L4_POLICY`) and only allows traffic if it is allowed at L3. Let me know if you'd like me to take a stab at this, but either way this refactoring should be a separate patch from removing `POLICY_INGRESS`/`POLICY_EGRESS`.

Reply:
@joestringer would be awesome if you could take that effort on. Would you like to do it as a commit within this branch, or an entirely separate patch?

Reply:
Can go either way. I don't expect it to be a huge patch and this PR isn't too big yet.

Reply:
Figured it might be better to put in a separate PR so we can do a full CI run on it and hopefully avoid forcing you to triage any issues that might arise with it: https://github.com/cilium/cilium/pull/5321